source/MaterialXCross/Cross.cpp

```diff
@@ -0,0 +1,286 @@
+    // easiest workaround for this is to provide a dummy entry point.
+    static const std::string dummyMain =
+        "void main()\n"
+        "{\n"
+        "}\n\n";
+
+    const char* shaderStrings[]{
+        glslPrivateUniforms.data(),
+        glslFragment.data(),
+        dummyMain.data()
+    };
+
+    const int stringLengths[]{
+        static_cast<int>(glslPrivateUniforms.size()),
+        static_cast<int>(glslFragment.size()),
+        static_cast<int>(dummyMain.size())
+    };
+
+    glslang::TShader shader(EShLangFragment);
+    shader.setStringsWithLengths(shaderStrings, stringLengths, 3);
+    shader.setEntryPoint("main");
+
+    // Permits uniforms without explicit layout locations
+    shader.setAutoMapLocations(true);
+
+    constexpr auto messages = static_cast<EShMessages>(
+        EShMsgSpvRules | EShMsgKeepUncalled | EShMsgDebugInfo
+    );
+
+    {
+        // The MaterialX GLSL generator is hardcoded to output GLSL 4.0
+        constexpr int defaultVersion = 400;
+        constexpr bool forwardCompatible = false;
+        
+        // Don't support includes in the GLSL source.
+        glslang::TShader::ForbidIncluder forbidIncluder;
+
+        if (!shader.parse(
+            &defaultTBuiltInResource,
+            defaultVersion,
+            forwardCompatible,
+            messages,
+            forbidIncluder
+        ))
+        {
+            const char* const log = shader.getInfoLog();
+            throw Exception(
+                std::string("glslang failed to parse the GLSL fragment:\n") + log
+            );
+        }
+    }
+
+    glslang::TProgram program;
+    program.addShader(&shader);
+    if (!program.link(messages))
+    {
+        const char* const log = program.getInfoLog();
+        throw Exception(
+            std::string("glslang failed to link the GLSL fragment:\n") + log
+        );
+    }
+
+    std::vector<uint32_t> spirv;
+
+    {
+        glslang::SpvOptions options;
+        options.generateDebugInfo = true;   // necessary to preserve symbols
+        options.disableOptimizer = true; 
+        options.optimizeSize = false;
+
+        glslang::GlslangToSpv(
+            *program.getIntermediate(EShLangFragment), spirv, &options
+        );
+    }
+
+    return spirv;
+}
+
+class HlslFragmentCrossCompiler : public spirv_cross::CompilerHLSL
+{
+public:
+    using spirv_cross::CompilerHLSL::CompilerHLSL;
+
+    /// Override this method to omit all uniforms except light data in the HLSL
+    /// fragment. VP2 will generate these uniforms in the final HLSL shader
+    /// based on XML wrapper properties, so emitting these would lead to double
+    /// definition compilation errors.
+    /// Please note that the base class method had to be made protected in our
+    /// SPIRV-Cross fork to be able to delegate to it. So we need to either
+    /// maintain our fork (definitely for short-term) or contribute the change
+    /// upstream.
```

Comment:
Actually, we won't need this when we implement proper Maya lighting support and don't need to emit the light data uniform.